src/Arcus.WebApi.Correlation/CorrelationMiddleware.cs

```diff
@@ -0,0 +1,131 @@
+        /// <returns>A <see cref="T:System.Threading.Tasks.Task" /> that represents the execution of this middleware.</returns>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            Guard.NotNull(httpContext, nameof(httpContext));
+            Guard.For<ArgumentException>(() => httpContext.Response is null, "Requires a 'Response'");
+            Guard.For<ArgumentException>(() => httpContext.Response.Headers is null, "Requires a 'Response' object with headers");
+
+            if (httpContext.Request.Headers.TryGetValue(_options.Transaction.HeaderName, out StringValues transactionIds))
+            {
+                if (!_options.Transaction.AllowInRequest)
+                {
+                    _logger.LogError("No correlation request header '{HeaderName}' for transaction ID was allowed in request", _options.Transaction.HeaderName);
+                    httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    await httpContext.Response.WriteAsync($"No correlation transaction ID request header '{_options.Transaction.HeaderName}' was allowed in the request");
+
+                    return;
+                }
+
+                _logger.LogTrace("Correlation request header '{HeaderName}' found with transaction ID '{TransactionId}'", _options.Transaction.HeaderName, transactionIds);
+            }
+
+            (string operationId, string transactionId) = DetermineCorrelationIds(httpContext, transactionIds);
+            var correlation = new CorrelationFeature(operationId, transactionId);
+            httpContext.Features.Set(correlation);
+
+            AddCorrelationResponseHeaders(httpContext, operationId, transactionId);
+
+            // TODO: on exception, should we enrich the exception.Data with the correlation info?
+            await _next(httpContext);
+        }
+
+        private (string operationId, string transactionId) DetermineCorrelationIds(HttpContext httpContext, StringValues transactionIds)
+        {
+            // TODO: make ID generation configurable for the consumer.
+            string operationId = httpContext.TraceIdentifier ?? Guid.NewGuid().ToString();
+            
+            string transactionId;
+            if (String.IsNullOrWhiteSpace(transactionIds.ToString()))
+            {
+                if (_options.Transaction.GenerateWhenNotSpecified)
+                {
+                    transactionId = Guid.NewGuid().ToString();
+                }
+                else
+                {
+                    transactionId = null;
+                }
+            }
+            else
+            {
+                transactionId = transactionIds.ToString();
+            }
+
+            return (operationId, transactionId);
+        }
+
+        private void AddCorrelationResponseHeaders(HttpContext httpContext, string operationId, string transactionId)
+        {
+            if (_options.Operation.IncludeInResponse)
```

Comment:
Can't we consolidate this as following?

```csharp
httpContext.Response.OnStarting(() =>
{
    if (_options.Operation.IncludeInResponse)
    {
        TryAddResponseHeader(httpContext, _options.Operation.HeaderName, operationId);
    }
    if (_options.Transaction.IncludeInResponse)
    {
        TryAddResponseHeader(httpContext, _options.Transaction.HeaderName, transactionId);
    }

    return Task.CompletedTask;
});
```

Reply:
Yeah was also thinking, but I don't know the lifecycle of the options instance, that's why I moved it outside the lambda so we don't get capture-variable problems.

Reply:
That's a good point, but how it is now, do we actually return the operation id or not? Because we are calling `OnStarting` twice which feels a bit odd

Reply:
Yeah; both are added. The `OnStarting` _Adds_ a delegate.